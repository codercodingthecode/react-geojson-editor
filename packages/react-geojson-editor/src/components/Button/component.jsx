import React from 'react'; // eslint-disable-line import/no-extraneous-dependencies
import PropTypes from 'prop-types'; // eslint-disable-line import/no-extraneous-dependencies

import ButtonStyles from './styles.scss';

export const Button = ({ children, ...props }) => (
  <button
    className={ButtonStyles.button}
    {...props}
  >
    {children}
  </button>
);

Button.propTypes = {
  children: PropTypes.string,
};

Button.defaultProps = {
  children: 'Button',
};

export default Button;
