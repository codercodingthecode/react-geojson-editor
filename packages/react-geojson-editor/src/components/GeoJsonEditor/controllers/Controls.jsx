// @ts-check
import React, { Component } from 'react';
import { GoogleMapContext } from '../google-map';

/**
 * React version of this Google sample Example
 * source: https://developers.google.com/maps/documentation/javascript/examples/control-custom
 */
export default class Controls extends Component {
    static contextType = GoogleMapContext;

    // @TODO this.controlId = IDGenerator.next();
    gmapControls = null;
    gmapControl = null;
    controlId = `gmap-control-${Math.random().toFixed(5).replace('0.', '')}`;
    controlRef = React.createRef();

    componentDidMount() {
      const { map, google } = this.context;
      this.gmapControls = map
        .controls[google.maps.ControlPosition.TOP_RIGHT];
      this.gmapControl = this.controlRef.current.firstChild;

      // console.log('[Controls centerControlDiv] 1', this.controlElement, this.controlRef.current.firstChild);

      this.gmapControls.push(this.gmapControl);
      // console.log('[Controls centerControlDiv] 1', centerControlDiv)
    }

    componentWillUnmount() {
      const controlIndex = this.gmapControls.indexOf(this.gmapControl);
      if (controlIndex >= 0) {
        this.gmapControls.removeAt(controlIndex);
      }
    }

    render() {
      const { children } = this.props;
      return (<div ref={this.controlRef} style={{ display: 'none' }}>
        <div id={this.controlId}>
          <ControlUI>
            {children}
          </ControlUI>
        </div>
              </div>);
    }
}

const controlUIStyle = {
  backgroundColor: '#fff',
  // border: '2px solid #fff',
  border: 'none',
  // borderRadius: '2px',
  borderRadius: '0',
  // boxShadow: '0 2px 6px rgba(0,0,0,.3)',
  boxShadow: 'none',
  cursor: 'pointer',
  marginTop: '10px',
  marginRight: '10px',
  marginBottom: '22px',
};

export function ControlUI({ children }) {
  return (<div style={{ ...controlUIStyle, textAlign: 'center' }}>
    {children}
  </div>);
}
