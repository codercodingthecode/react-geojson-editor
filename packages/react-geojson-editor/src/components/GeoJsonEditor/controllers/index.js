export { default as GeoJsonController } from './GeoJsonController';
export { default as PolygonController } from './PolygonController';

