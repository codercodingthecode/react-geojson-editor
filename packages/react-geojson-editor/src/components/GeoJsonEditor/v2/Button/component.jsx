import React, { Component, Children } from 'react';

import ButtonStyles from './styles.scss';

export const Button = ({ children, ...props }) => (
  <button
    className={ButtonStyles.button}
    {...props}
  >{children}
  </button>
);

export default Button;
