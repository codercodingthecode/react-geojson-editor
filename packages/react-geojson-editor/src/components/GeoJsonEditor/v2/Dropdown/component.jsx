import React, { Component, Children } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

// import OutsideClickHandler from 'react-outside-click-handler';
import DropdownStyles from './styles.scss';
import { StylesType } from '../../../../custom-types';

export function renderContent(children) {
  return (
    <div className={DropdownStyles.dropdownBody}>
      {children}
    </div>
  );
}

export class Dropdown extends Component {
  constructor(props) {
    super(props);
    this.onBlurClose = this.onBlurClose.bind(this);
    this.toggleOpen = this.toggleOpen.bind(this);
    // this.renderContent = this.renderContent.bind(this);
  }
    state = {
      open: this.props.openAtStart,
    }

    onBlurClose() {
      this.setState({ open: false });
    }

    toggleOpen(e) {
      e.preventDefault();
      const { children } = this.props;
      this.setState({ open: !this.state.open && Boolean(Children.count(children)) });
    }

    render() {
      const { open } = this.state;
      const {
        children, name, positionRight, style, containerStyle,
      } = this.props;
      return (
        <div
          className={
                positionRight ?
                    classnames(DropdownStyles.mlDropdown, DropdownStyles.rightPosition) :
                    DropdownStyles.mlDropdown
            }
          style={style}
        >
          <div className={DropdownStyles.dropdownContainer} style={containerStyle}>
            {/* <OutsideClickHandler onOutsideClick={this.onBlurClose}> */}
            <button
              className={
                open ?
                  classnames(DropdownStyles.dropdownButton, DropdownStyles.open) :
                  DropdownStyles.dropdownButton
              }
              onClick={this.toggleOpen}
            >
              {name}
            </button>
            {open && renderContent(children)}
            {/* </OutsideClickHandler> */}
          </div>
        </div>
      );
    }
}

Dropdown.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.func,
    PropTypes.arrayOf(PropTypes.node),
  ]).isRequired,
  name: PropTypes.string,
  style: StylesType,
  containerStyle: StylesType,
  positionRight: PropTypes.bool,
  openAtStart: PropTypes.bool,
};

Dropdown.defaultProps = {
  name: 'Dropdown',
  style: {},
  containerStyle: {},
  positionRight: false,
  openAtStart: false,
};

export default Dropdown;
