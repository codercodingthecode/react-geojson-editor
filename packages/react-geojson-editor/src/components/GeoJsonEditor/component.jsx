import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { GoogleMapWithLoader, PolygonDataProvider } from './google-map';
// import Reader from './geojson-manager/Reader';
import { GeoJsonEditorMode } from '../../utils';
import { GeoJsonType } from '../../custom-types';
import { PolygonController, GeoJsonController } from './controllers';
import { GeoJsonWriter } from './writer';
import { GeoJsonReader } from './reader';

export class GeoJsonEditor extends Component {
    state = {
      controlMode: GeoJsonEditorMode.VIEW,
    };

    render() {
      const {
        center, zoom, onSave, onDelete, existingPolygons, mapHeight, mapWidth,
      } = this.props;
      const { controlMode } = this.state;
      return (
        <div style={{ position: 'relative' }}>
          <GoogleMapWithLoader
            googleMapURL={`https://maps.googleapis.com/maps/api/js?key=${this.props.googleMapKey}&v=3.exp&libraries=geometry,drawing,places`}
            center={center}
            zoom={zoom}
            height={mapHeight}
            width={mapWidth}
          >
            <PolygonDataProvider
              initialGeojson={existingPolygons}
              onSave={onSave}
              onDelete={onDelete}
            >

              <GeoJsonController
                mode={controlMode}
                onToggleMode={mode => this.setState({ controlMode: mode })}
              />
              {controlMode === GeoJsonEditorMode.EDIT && <PolygonController />}

              {controlMode === GeoJsonEditorMode.EDIT ? <GeoJsonWriter /> : <GeoJsonReader />}

            </PolygonDataProvider>
          </GoogleMapWithLoader>
        </div>
      );
    }
}

export default GeoJsonEditor;

GeoJsonEditor.propTypes = {
  center: PropTypes.shape({
    lng: PropTypes.number,
    lat: PropTypes.number,
  }),
  zoom: PropTypes.number,
  googleMapKey: PropTypes.string,
  existingPolygons: GeoJsonType,
  onDelete: PropTypes.func,
  onSave: PropTypes.func,
  //   onUpdate: PropTypes.func,
  mapHeight: PropTypes.number,
  mapWidth: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};
GeoJsonEditor.defaultProps = {
  center: { lng: -73.945, lat: 40.674 },
  zoom: 5,
  existingPolygons: null,
  googleMapKey: process.env.GOOGLE_MAP_KEY,
  mapHeight: 500,
  mapWidth: '100%',
  onDelete: () => {},
  onSave: () => {},
};
