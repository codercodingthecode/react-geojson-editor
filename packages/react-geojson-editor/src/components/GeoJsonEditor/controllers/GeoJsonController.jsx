import React from 'react';
import PropTypes from 'prop-types';
import { GeoJsonEditorMode } from '../../../utils';
import { Dropdown } from '../../Dropdown';
import { Button } from '../../Button';
import { PolygonDataContext } from '../google-map'; // @TODO
import Controls from './Controls';

/**
 * - show ui for enabling polygon editing capability
 * - dispatch actions for
 *      - persting new changes
 *      - remove new changes
 *      - toggle edit mode
 * @param {*} props
 */
export const GeoJsonController = ({
  mode, onToggleMode,
}) => (
  <Controls title="Managing geojson changes">
    {mode !== GeoJsonEditorMode.VIEW ?
      <Dropdown name="Changes" openAtStart style={{ padding: 0 }} containerStyle={{ width: 120 }} >
        <PolygonDataContext.Consumer>
          {({ discardChanges, persistChanges }) =>
                        (
                          <div style={{ padding: 0 }} title="GeoJsonControls">
                            <Button
                              onClick={(ev) => {
                                        persistChanges(
                                            ev,
                                            () => onToggleMode(GeoJsonEditorMode.VIEW),
                                        );
                                    }}
                              title="Save changes"
                            >
                                    Save
                            </Button>
                            <Button
                              onClick={ev => discardChanges(
                                        ev,
                                        () => onToggleMode(GeoJsonEditorMode.VIEW),
                                    )
                                    }
                              title="Discard changes"
                            >
                                    Cancel
                            </Button>
                          </div>
                        )
                    }
        </PolygonDataContext.Consumer>
      </Dropdown> :
      <Button
        onClick={() => onToggleMode(GeoJsonEditorMode.EDIT)}
        style={{
            width: 120,
            boxShadow: 'rgba(0, 0, 0, 0.3) 0px 2px 6px',
            borderRadius: 2,
        }}
        title="Turn on Edit mode"
      >
                Create / Edit
      </Button>
        }
  </Controls>
);
export default GeoJsonController;

GeoJsonController.propTypes = {
  mode: PropTypes.string,
  onToggleMode: PropTypes.func,
};

GeoJsonController.defaultProps = {
  mode: GeoJsonEditorMode.VIEW,
  onToggleMode: () => { },
};
