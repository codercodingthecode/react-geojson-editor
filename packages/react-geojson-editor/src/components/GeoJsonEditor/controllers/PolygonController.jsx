import React from 'react';
// import PropTypes from 'prop-types';
// import Reader from './geojson-manager/Reader';

// import { GeoJsonEditorMode } from '../../../../utils';
import { Dropdown } from '../../Dropdown';
import { Button } from '../../Button';
import Controls from './Controls';
// import { PolygonDataContext } from '../GoogleMap'; // @TODO
import { PolygonDataContext } from '../google-map'; // @TODO

/**
 * - show ui for enabling polygon mutation operations
 */
export const PolygonController = () => (
  <Controls title="Managing polygon edits">
    <PolygonDataContext.Consumer>
      {({ selectedPolygonList, subtractAction, deleteAction }) => {
        const selectedLength = selectedPolygonList.length;
        return (
          <Dropdown name={`${selectedLength} Selected`} style={{ padding: 0 }} containerStyle={{ width: 120 }} >
            {/* @TODO put this code inside its EditorControls component */}
            <div style={{ padding: 0 }} title="EditorControls">
              {/* <Button disabled={selectedLength < 2} title="Merge">Add TODO</Button> */}
              <Button
                disabled={selectedLength < 2}
                onClick={subtractAction}
                title="Make Hollow"
              >
                Subtract
              </Button>
              <Button
                disabled={selectedLength <= 0}
                onClick={deleteAction}
              >
                Delete
              </Button>
            </div>
          </Dropdown>
        );
      }}
    </PolygonDataContext.Consumer>
  </Controls>
);

export default PolygonController;
