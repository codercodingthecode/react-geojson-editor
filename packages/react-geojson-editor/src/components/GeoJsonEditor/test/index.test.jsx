import React from 'react';

import { shallow } from 'enzyme';
import { GeoJsonEditor } from '../component';
import { GeoJsonController } from '../controllers';

describe('Button', () => {
  it('has not changed', () => {
    const wrapper = shallow(<GeoJsonEditor />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders', () => {
    const wrapper = shallow(<GeoJsonEditor />);
    expect(wrapper.find(GeoJsonController)).toHaveLength(1);
  });
});
