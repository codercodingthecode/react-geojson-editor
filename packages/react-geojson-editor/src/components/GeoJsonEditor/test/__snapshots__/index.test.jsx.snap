// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has not changed 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GeoJsonEditor
    center={
      Object {
        "lat": 40.674,
        "lng": -73.945,
      }
    }
    existingPolygons={null}
    googleMapKey={undefined}
    mapHeight={500}
    mapWidth="100%"
    onDelete={[Function]}
    onSave={[Function]}
    zoom={5}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <GoogleMapWithLoader
        center={
          Object {
            "lat": 40.674,
            "lng": -73.945,
          }
        }
        googleMapURL="https://maps.googleapis.com/maps/api/js?key=undefined&v=3.exp&libraries=geometry,drawing,places"
        height={500}
        width="100%"
        zoom={5}
      >
        <PolygonDataProvider
          initialGeojson={null}
          onDelete={[Function]}
          onSave={[Function]}
        >
          <Unknown
            mode="VIEW"
            onToggleMode={[Function]}
          />
          <Unknown />
        </PolygonDataProvider>
      </GoogleMapWithLoader>,
      "style": Object {
        "position": "relative",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "center": Object {
          "lat": 40.674,
          "lng": -73.945,
        },
        "children": <PolygonDataProvider
          initialGeojson={null}
          onDelete={[Function]}
          onSave={[Function]}
        >
          <Unknown
            mode="VIEW"
            onToggleMode={[Function]}
          />
          <Unknown />
        </PolygonDataProvider>,
        "googleMapURL": "https://maps.googleapis.com/maps/api/js?key=undefined&v=3.exp&libraries=geometry,drawing,places",
        "height": 500,
        "width": "100%",
        "zoom": 5,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Unknown
              mode="VIEW"
              onToggleMode={[Function]}
            />,
            false,
            <Unknown />,
          ],
          "initialGeojson": null,
          "onDelete": [Function],
          "onSave": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "mode": "VIEW",
              "onToggleMode": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <GoogleMapWithLoader
          center={
            Object {
              "lat": 40.674,
              "lng": -73.945,
            }
          }
          googleMapURL="https://maps.googleapis.com/maps/api/js?key=undefined&v=3.exp&libraries=geometry,drawing,places"
          height={500}
          width="100%"
          zoom={5}
        >
          <PolygonDataProvider
            initialGeojson={null}
            onDelete={[Function]}
            onSave={[Function]}
          >
            <Unknown
              mode="VIEW"
              onToggleMode={[Function]}
            />
            <Unknown />
          </PolygonDataProvider>
        </GoogleMapWithLoader>,
        "style": Object {
          "position": "relative",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "center": Object {
            "lat": 40.674,
            "lng": -73.945,
          },
          "children": <PolygonDataProvider
            initialGeojson={null}
            onDelete={[Function]}
            onSave={[Function]}
          >
            <Unknown
              mode="VIEW"
              onToggleMode={[Function]}
            />
            <Unknown />
          </PolygonDataProvider>,
          "googleMapURL": "https://maps.googleapis.com/maps/api/js?key=undefined&v=3.exp&libraries=geometry,drawing,places",
          "height": 500,
          "width": "100%",
          "zoom": 5,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Unknown
                mode="VIEW"
                onToggleMode={[Function]}
              />,
              false,
              <Unknown />,
            ],
            "initialGeojson": null,
            "onDelete": [Function],
            "onSave": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "mode": "VIEW",
                "onToggleMode": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
